### PR TITLE
[FLINK-21480][runtime] Respect external resources from resource requirements

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Respect external resources from resource requirements

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by xintongsong — COMMENTED** [review]
Thanks for opening the PR, @KarmaGYZ. I have left some inline comment.
In addition, I'd suggest to re-organize the commits, separating making external resources available in `xxxResourceSpec` and making it take effect in RM and its drivers.

**Comment by KarmaGYZ** [comment]
@xintongsong Thanks for the review! PR updated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorProcessSpec.java
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorProcessSpec.java:109]: Same for `WorkerResourceSpec` and `TaskExecutorResourceSpec`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/WorkerResourceSpec.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/WorkerResourceSpec.java:242]: Same here for existing external resource names.

**Review by xintongsong — COMMENTED** [review]
Thanks @KarmaGYZ. I have a few more comments. Please take a look.
In addition to the inline comment:
- nit: The component of the 3rd & 4th commits should be `[core]` rather than `[runtime]`, and the last should be `[k8s][yarn]` rather than `[runtime]`.
- It would be better to verify the changes with real fine-grained external resource requirements before merging.

**Comment by KarmaGYZ** [comment]
Thanks for the review @xintongsong . I've verified it in a Kubernetes cluster.

File: flink-core/src/main/java/org/apache/flink/api/common/resources/Resource.java
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/resources/Resource.java:91]: It would be nice to keep the scope of `@SuppressWarnings` as small as possible. In this case, the single statement `T other = (T) o;` should be enough.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: JavaDoc style is preferred. It hints the compiler/IDE that there's a binding between the docs and the following class/field/method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java:643]:
```
Add the given extended resource. The old value with the same resource name will be replaced if present.
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java:649]:
```
Add the given extended resources. This will discard all the previous added extended resources.
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: Same for `ResourceSpec.Builder` and `WorkerResourceSpec.Builder`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java:492]: What if there's no external resource? We will have something like `ResourceProfile{a=b, c=d, }`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorProcessSpec.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorProcessSpec.java:141]: `checkArgument` is preferred.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java:492]: Same for `TaskExecutorProcessSpec` and `WorkerResourceSpec`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/externalresource/ExternalResourceUtils.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/externalresource/ExternalResourceUtils.java:53]: I'd suggest to:
- Use the phrase `none`
- Add the constant in `ExternalResourceOptions`
- Mention in the description of `EXTERNAL_RESOURCE_LIST` that the keyword is preserved

File: flink-runtime/src/main/java/org/apache/flink/runtime/externalresource/ExternalResourceUtils.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/externalresource/ExternalResourceUtils.java:164]: `Map<String, ExternalResource>` -> `Collection<ExternalResource>`

File: flink-runtime/src/main/java/org/apache/flink/runtime/externalresource/ExternalResourceUtils.java
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/externalresource/ExternalResourceUtils.java:174]: This can be simplified as follows:
```
return extendedResources.entrySet().stream()
        .map(e -> e.getKey() + "=" + e.getValue().getValue())
        .collect(Collectors.joining(", "));
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorProcessSpec.java
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorProcessSpec.java:141]: Same for `WorkerResourceSpec` and `TaskExecutorResourceSpec`.

**Review by xintongsong — APPROVED** [review]
Thanks @KarmaGYZ. LGTM. I'll merge this after AZP gives green light.

**Pull request closed without merging**
